File: src/main/java/net/tlalka/imager/utils/RxUtils.java
package net.tlalka.imager.utils;

import io.reactivex.functions.BiConsumer;
import io.reactivex.functions.BiFunction;

public class RxUtils {

    private RxUtils() {
    }

    public static <T> BiFunction<T, T, T> comp(BiConsumer<T, T> func) {
        return (t1, t2) -> {
            func.accept(t1, t2);
            return t2;
        };
    }
}

File: src/main/kotlin/net/tlalka/imager/utils/RxUtils.kt
package net.tlalka.imager.utils

import io.reactivex.Observable
import io.reactivex.Single

object RxUtils {

    fun <T> Observable<T>.comp(func: (T, T) -> Unit): Observable<T> {
        return this.scan { t1: T, t2: T -> func(t1, t2); t2 }
    }

    fun <T, R> Observable<T>.flatSingle(single: (T) -> Single<R>): Observable<R> {
        return this.flatMap { t1 : T -> single(t1).toObservable() }
    }

    fun <T, R> Observable<T>.flatSuccess(func: (T) -> Observable<R>): Observable<R> {
        return this.flatMap { t1 : T -> func(t1).onErrorResumeNext(Observable.empty()) }
    }
}
